File: Aho/infix_to_postfix/infix_to_postfix.js
import * as process from "process"
import {exprString_c} from "./expression_string.js"

/**@type {string} */
let lookahead
let expressionLine = new exprString_c("9+5-2")
let expressionOut = new exprString_c("")

function expr() {
    term()
    while (true) {
        if (lookahead == "+") {
            match("+")
            term()
            expressionOut.appendChar("+")
        } else if (lookahead == "-") {
            match("-")
            term()
            expressionOut.appendChar("-")
        } else {
            return
        }
    }
}

function term() {
    if (!isNaN(Number(lookahead))) {
        let t = lookahead
        match(lookahead)
        expressionOut.appendChar(t)
    } else {
        console.log("syntax error in term()! " + lookahead + "- is not number!")
        process.exit(0)
    }
}

/**
 * @param {string} t
 */
function match(t) {
    if (lookahead == t) {
        [, lookahead] = expressionLine.getNextChar()
    } else {
        console.log("syntax error in match()!")
        process.exit(0)
    }
}

(function main() {
    [, lookahead] = expressionLine.getNextChar()

    expr()

    console.log(expressionOut.expression)
}) ()

File: Aho/infix_to_postfix/infix_to_postfix.c

#include <stdio.h>
#include <ctype.h>
#include <stdint.h>
#include <stdbool.h>

#include "expression_string.h"

expressionString_s foo;
int32_t lookahead = 0;

void expr() {
    term();
    while (true) {
        if (lookahead == '+') {
            match('+');
            term();
            printf("+");
        } else if (lookahead == '-') {
            match('-');
            term();
            printf("-");
        } else {
            return;
        }
    }
}

void term() {
        if (isdigit(lookahead)) {
            char t = lookahead;
            match(lookahead);
            printf("%c", t);
        } else {
            printf("syntax error in term()! %c - is not number!\n", (lookahead));
            exit(0);
        }
}

void match(int32_t t) {
    if (lookahead == (char)t) {
        lookahead = getNextChar(&foo);
    } else {
        printf("Syntax error in match!\n");
        exit(0);
    }
}

int main(int argc, char **argv) {
    foo = buildExpressionString("9+2-5");
    printf("%s\n", foo.string);

    lookahead = getNextChar(&foo);
    expr();
    printf("\n");
}